LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -30,96 +36,103 @@ internal sealed class ConcentratorDeduplication<T> :
+            var stationEui = loRaRequest.StationEui;
 
             StationEui previousStation;
             lock (this.cache)
             {
                 if (!this.cache.TryGetValue(key, out previousStation))
                 {
-                    AddToCache(key, stationEui);
-                    return false;
+                    _ = this.cache.Set(key, stationEui, new MemoryCacheEntryOptions()
+                    {
+                        SlidingExpiration = DefaultExpiration
+                    });
+                    return Result.NotDuplicate;
                 }
             }
 
-            if (previousStation == stationEui)
+            if (LoRaRequest.RequiresConfirmation(loRaRequest) && previousStation == stationEui)
             {
-                // considered as a resubmit
-                this.logger.LogDebug("Message received from the same EUI {StationEui} as before, will not drop.", stationEui);
-                return false;
+                this.logger.LogDebug($"Message received from the same EUI {stationEui} as before, will not drop.");
+                return Result.DuplicateDueToResubmission;
             }
 
             // received from a different station
-            if (this.socketRegistry.IsSocketWriterOpen(previousStation))
+            if (ShouldDrop(loRaRequest, loRaDevice))
             {
                 this.logger.LogInformation($"{Constants.DuplicateMessageFromAnotherStationMsg} with EUI {stationEui}, will drop.");
-                return true;
+                return Result.Duplicate;
             }
 
-            this.logger.LogInformation("Connectivity to previous station with EUI {PreviousStation}, was lost, will not drop and will use station with EUI {StationEui} from now onwards.",
-                                       previousStation, stationEui);
-            AddToCache(key, stationEui);
-            return false;
+            this.logger.LogDebug($"Message from station with EUI {stationEui} will not be dropped due to DeduplicationStrategy.");
+            return Result.SoftDuplicate;
         }
 
-        internal static string CreateCacheKey(T frame)
-            => frame switch
+        internal static string CreateCacheKey(LoRaRequest loRaRequest)
+            => loRaRequest.Payload switch
             {
-                UpstreamDataFrame asDataFrame => CreateCacheKeyCore(asDataFrame),
-                JoinRequestFrame asJoinRequestFrame => CreateCacheKeyCore(asJoinRequestFrame),
-                _ => throw new ArgumentException($"{frame} with invalid type.")
+                LoRaPayloadData asDataPayload => CreateCacheKeyCore(asDataPayload),
+                LoRaPayloadJoinRequest asJoinPayload => CreateCacheKeyCore(asJoinPayload),
+                _ => throw new ArgumentException($"{loRaRequest} with invalid type.")
             };
 
-        private static string CreateCacheKeyCore(UpstreamDataFrame updf)
+        private static string CreateCacheKeyCore(LoRaPayloadData payload)
```

Comment:
can we make those unit testable? I feel like the logic in here, needs to be tested thoroughly.

Reply:
Right now we unit test whether we get the expected output. I think we discussed with Atif whether we should test further but dropped the idea as this would be an implementation detail bound to change (might remember wrong though)? This will be the case actually if we simplify keys in #1003 because now we have the DevEUI - if ok for you I will address this in that task.

Since we are using static methods within not sure what we can unit test further. Also the output is hashed so we can't observe whether a field is in there either - we can only check if the output changed.

Reply:
What I was thinking is that we should test that the key actually changes when we change the relevant input parameter from the parameters. We also have a few variants with optional arguments (RawMessage) that may or may not be included and impacts the key generation. Even if the implementation may change, it would add value having tests for the current implementation. The linked item is not assigned yet and could be dropped as it is a nice to have.

There are multiple options on how you can make this testable. If you want to keep it static, you can mark it protected, so you can call it from a test class that inherits from the ConcentratorDeduplication. If you want to validate not only the hash but also the content, you could factor out the hash computation into a virtual protected method you can override that you could mock and not create a hash.

Reply:
Created #1003 to keep track of this.